lib/Analysis/AxisInfo.cpp

```diff
@@ -301,10 +301,19 @@ class MulIOpAxisInfoVisitor final : public BinaryOpVisitorImpl<arith::MulIOp> {
 
   int64_t getDivisibility(arith::MulIOp op, const AxisInfo &lhs,
                           const AxisInfo &rhs, int dim) override {
-    // lhs = k * d_lhs
-    // rhs = p * d_rhs
-    // lhs * rhs = k * d_lhs * p * d_rhs = k * p * d_lhs * d_rhs
-    return lhs.getDivisibility(dim) * rhs.getDivisibility(dim);
+    auto lhsDivisibility = lhs.getDivisibility(dim);
+    if (lhs.getContiguity(dim) > 1 &&
+        !(rhs.getConstantValue().has_value() && rhs.getConstantValue() == 1)) {
+      // Ignore the 'fake' divisibility
```

Comment:
As my understanding:
- [0,1,2,3,8,9,10,11,4,5,6,7,...] contiguity is 4
- [0,0,0,0,1,1,1,1,2,2,2,2,...] constancy is 4
- [0,8,4,12,16,24,...] divisibility is 4
But the op MakeRange doesn't follow the rule above! The code in MakeRangeOpAxisInfoVisitor tells that it may create a vector [16,17,18,19] and set contiguity=4, constancy=1, divisibility=16 (instead of 1). So I understand that the divisibility is not simply "What is the value multiple of", also there is a special case: When the dim is contiguous, the contiguous values are not counted, only the start value is used to calculate the (fake) divisibility. But after the vector is MUL with some none-1 value, the vector is not contiguous any more. Think about [16,17,18,19]*2 => [32,34,36,38]. Here, the 'true' divisibility 1 should be used to calculate the new divisibility: 1*2=2 (the correct divisibility of [32,34,36,38]).

Reply:
> a vector [16,17,18,19] and set contiguity=4, constancy=1, divisibility=16 (instead of 1).

I still think the divisibility should be 16. As it's defined, divisibility is the largest power of two of the leading element in the contiguous group. The contiguous group here is 16 but not 1.

If you have [2,3, 8,9] the divisibility would be 2 but not 1

Reply:
The calculation seems correct to me but I don't think 16 is "fake" divisibility

Reply:
Yep, I agree with Keren!

Reply:
OK, maybe my term 'fake' is not accurate. What's a better term to call this? Let me sum up the logic in this PR:

- After a MulI, ShLI or ShR, a contiguous group will no longer be contiguous, e.g. [16,17,18,19] x 2 => [32,34,36,38], so we should use 1 (instead of 16) to calculate the new divisibility. The only exception is MulI 1, or shift 0.
- For ExpandDims, [16,17,18,19] may become [[16],[17],[18],[19]], and with further more OPs, maybe something like [[16,16],[17,17],[18,18],[19,19]]. We can see the divisibility along the new dim should also be calculated using 1 (instead of 16).

Reply:
I would suggest just remove the comment of "fake" and leave an example here.

Reply:
> I would suggest just remove the comment of "fake" and leave an example here.

Changed in the lattest commit.